File: src/aws/lambdas/incremental_distribution/src/test/java/uk/nhs/nhsx/diagnosiskeydist/DistributionServiceWindowTest.java
package uk.nhs.nhsx.diagnosiskeydist;

import org.junit.jupiter.api.Test;

import java.time.LocalDateTime;
import java.time.ZoneId;
import java.util.Calendar;
import java.util.Date;
import java.util.TimeZone;

import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

public class DistributionServiceWindowTest {
    @Test
    public void testIsDistributionTimeWindowStartInvalid() {
        DistributionServiceWindow earlyStartDate = new DistributionServiceWindow(utcDate(2020, 7, 4, 11, 45, 59, 999));
        DistributionServiceWindow lateStartDate = new DistributionServiceWindow(utcDate(2020, 7, 4, 11, 48, 0, 0));
        DistributionServiceWindow evenHourEarlyStartDate = new DistributionServiceWindow(utcDate(2020, 7, 4, 12, 46, 0, 0));
        DistributionServiceWindow evenHourLateStartDate = new DistributionServiceWindow(utcDate(2020, 7, 4, 12, 47, 0, 0));
        assertFalse(earlyStartDate.validBatchStartDate());
        assertFalse(lateStartDate.validBatchStartDate());
        assertFalse(evenHourEarlyStartDate.validBatchStartDate());
        assertFalse(evenHourLateStartDate.validBatchStartDate());
    }

    @Test
    public void testIsDistributionTimeWindowStartValid() {
    	DistributionServiceWindow earlyStartDate = new DistributionServiceWindow(utcDate(2020, 7, 4, 11, 46, 0, 0));
        DistributionServiceWindow lateStartDate = new DistributionServiceWindow(utcDate(2020, 7, 4, 11, 47, 59, 999));
        assertTrue(earlyStartDate.validBatchStartDate());
        assertTrue(lateStartDate.validBatchStartDate());
    }

    private Date utcDate(int year, int month, int day, int hour, int minute, int second, int millis) {
        Calendar cal = Calendar.getInstance(TimeZone.getTimeZone("UTC"));
        cal.set(Calendar.YEAR, year);
        cal.set(Calendar.MONTH, month - 1);
        cal.set(Calendar.DAY_OF_MONTH, day);
        cal.set(Calendar.HOUR_OF_DAY, hour);
        cal.set(Calendar.MINUTE, minute);
        cal.set(Calendar.SECOND, second);
        cal.set(Calendar.MILLISECOND, millis);

        return cal.getTime();
    }

    @Test
    public void nextFullHour() {
        assertThat(
                new DistributionServiceWindow(
                        utcDate(2020, 7, 4, 0, 0, 0, 0)
                ).nextFullHour().toInstant().atZone(ZoneId.of("UTC")).toLocalDateTime()
        ).isEqualTo(LocalDateTime.of(2020, 7, 4, 2, 0, 0, 0));

        assertThat(
                new DistributionServiceWindow(
                        utcDate(2020, 7, 4, 1, 0, 0, 0)
                ).nextFullHour().toInstant().atZone(ZoneId.of("UTC")).toLocalDateTime()
        ).isEqualTo(LocalDateTime.of(2020, 7, 4, 2, 0, 0, 0));

        assertThat(
                new DistributionServiceWindow(
                        utcDate(2020, 7, 4, 1, 59, 59, 999)
                ).nextFullHour().toInstant().atZone(ZoneId.of("UTC")).toLocalDateTime()
        ).isEqualTo(LocalDateTime.of(2020, 7, 4, 2, 0, 0, 0));

        assertThat(
                new DistributionServiceWindow(
                        utcDate(2020, 7, 4, 22, 0, 0, 0)
                ).nextFullHour().toInstant().atZone(ZoneId.of("UTC")).toLocalDateTime()
        ).isEqualTo(LocalDateTime.of(2020, 7, 5, 0, 0, 0, 0));

        assertThat(
                new DistributionServiceWindow(
                        utcDate(2020, 7, 4, 23, 0, 0, 0)
                ).nextFullHour().toInstant().atZone(ZoneId.of("UTC")).toLocalDateTime()
        ).isEqualTo(LocalDateTime.of(2020, 7, 5, 0, 0, 0, 0));

        assertThat(
                new DistributionServiceWindow(
                        utcDate(2020, 7, 4, 23, 59, 59, 999)
                ).nextFullHour().toInstant().atZone(ZoneId.of("UTC")).toLocalDateTime()
        ).isEqualTo(LocalDateTime.of(2020, 7, 5, 0, 0, 0, 0));
    }

    @Test
    public void zipExpirationExclusive() {
        assertThat(
                new DistributionServiceWindow(
                        utcDate(2020, 7, 4, 0, 0, 0, 0)
                ).zipExpirationExclusive().toInstant().atZone(ZoneId.of("UTC")).toLocalDateTime()
        ).isEqualTo(LocalDateTime.of(2020, 7, 4, 4, 0, 0, 0));

        assertThat(
                new DistributionServiceWindow(
                        utcDate(2020, 7, 4, 1, 0, 0, 0)
                ).zipExpirationExclusive().toInstant().atZone(ZoneId.of("UTC")).toLocalDateTime()
        ).isEqualTo(LocalDateTime.of(2020, 7, 4, 4, 0, 0, 0));

        assertThat(
                new DistributionServiceWindow(
                        utcDate(2020, 7, 4, 1, 59, 59, 999)
                ).zipExpirationExclusive().toInstant().atZone(ZoneId.of("UTC")).toLocalDateTime()
        ).isEqualTo(LocalDateTime.of(2020, 7, 4, 4, 0, 0, 0));

        assertThat(
                new DistributionServiceWindow(
                        utcDate(2020, 7, 4, 22, 0, 0, 0)
                ).zipExpirationExclusive().toInstant().atZone(ZoneId.of("UTC")).toLocalDateTime()
        ).isEqualTo(LocalDateTime.of(2020, 7, 5, 2, 0, 0, 0));

        assertThat(
                new DistributionServiceWindow(
                        utcDate(2020, 7, 4, 23, 0, 0, 0)
                ).zipExpirationExclusive().toInstant().atZone(ZoneId.of("UTC")).toLocalDateTime()
        ).isEqualTo(LocalDateTime.of(2020, 7, 5, 2, 0, 0, 0));

        assertThat(
                new DistributionServiceWindow(
                        utcDate(2020, 7, 4, 23, 59, 59, 999)
                ).zipExpirationExclusive().toInstant().atZone(ZoneId.of("UTC")).toLocalDateTime()
        ).isEqualTo(LocalDateTime.of(2020, 7, 5, 2, 0, 0, 0));
    }

    @Test
    public void earliestBatchStartDateWithinHourInclusive() {
        assertThat(
                new DistributionServiceWindow(
                        utcDate(2020, 7, 4, 0, 0, 0, 0)
                ).earliestBatchStartDateWithinHourInclusive().toInstant().atZone(ZoneId.of("UTC")).toLocalDateTime()
        ).isEqualTo(LocalDateTime.of(2020, 7, 4, 1, 46, 0, 0));

        assertThat(
                new DistributionServiceWindow(
                        utcDate(2020, 7, 4, 1, 0, 0, 0)
                ).earliestBatchStartDateWithinHourInclusive().toInstant().atZone(ZoneId.of("UTC")).toLocalDateTime()
        ).isEqualTo(LocalDateTime.of(2020, 7, 4, 1, 46, 0, 0));

        assertThat(
                new DistributionServiceWindow(
                        utcDate(2020, 7, 4, 1, 59, 59, 999)
                ).earliestBatchStartDateWithinHourInclusive().toInstant().atZone(ZoneId.of("UTC")).toLocalDateTime()
        ).isEqualTo(LocalDateTime.of(2020, 7, 4, 1, 46, 0, 0));

        assertThat(
                new DistributionServiceWindow(
                        utcDate(2020, 7, 4, 22, 0, 0, 0)
                ).earliestBatchStartDateWithinHourInclusive().toInstant().atZone(ZoneId.of("UTC")).toLocalDateTime()
        ).isEqualTo(LocalDateTime.of(2020, 7, 4, 23, 46, 0, 0));

        assertThat(
                new DistributionServiceWindow(
                        utcDate(2020, 7, 4, 23, 0, 0, 0)
                ).earliestBatchStartDateWithinHourInclusive().toInstant().atZone(ZoneId.of("UTC")).toLocalDateTime()
        ).isEqualTo(LocalDateTime.of(2020, 7, 4, 23, 46, 0, 0));

        assertThat(
                new DistributionServiceWindow(
                        utcDate(2020, 7, 4, 23, 59, 59, 999)
                ).earliestBatchStartDateWithinHourInclusive().toInstant().atZone(ZoneId.of("UTC")).toLocalDateTime()
        ).isEqualTo(LocalDateTime.of(2020, 7, 4, 23, 46, 0, 0));
    }

    @Test
    public void latestBatchStartDateWithinHourExclusive() {
        assertThat(
                new DistributionServiceWindow(
                        utcDate(2020, 7, 4, 0, 0, 0, 0)
                ).latestBatchStartDateWithinHourExclusive().toInstant().atZone(ZoneId.of("UTC")).toLocalDateTime()
        ).isEqualTo(LocalDateTime.of(2020, 7, 4, 1, 48, 0, 0));

        assertThat(
                new DistributionServiceWindow(
                        utcDate(2020, 7, 4, 1, 0, 0, 0)
                ).latestBatchStartDateWithinHourExclusive().toInstant().atZone(ZoneId.of("UTC")).toLocalDateTime()
        ).isEqualTo(LocalDateTime.of(2020, 7, 4, 1, 48, 0, 0));

        assertThat(
                new DistributionServiceWindow(
                        utcDate(2020, 7, 4, 1, 59, 59, 999)
                ).latestBatchStartDateWithinHourExclusive().toInstant().atZone(ZoneId.of("UTC")).toLocalDateTime()
        ).isEqualTo(LocalDateTime.of(2020, 7, 4, 1, 48, 0, 0));

        assertThat(
                new DistributionServiceWindow(
                        utcDate(2020, 7, 4, 22, 0, 0, 0)
                ).latestBatchStartDateWithinHourExclusive().toInstant().atZone(ZoneId.of("UTC")).toLocalDateTime()
        ).isEqualTo(LocalDateTime.of(2020, 7, 4, 23, 48, 0, 0));

        assertThat(
                new DistributionServiceWindow(
                        utcDate(2020, 7, 4, 23, 0, 0, 0)
                ).latestBatchStartDateWithinHourExclusive().toInstant().atZone(ZoneId.of("UTC")).toLocalDateTime()
        ).isEqualTo(LocalDateTime.of(2020, 7, 4, 23, 48, 0, 0));

        assertThat(
                new DistributionServiceWindow(
                        utcDate(2020, 7, 4, 23, 59, 59, 999)
                ).latestBatchStartDateWithinHourExclusive().toInstant().atZone(ZoneId.of("UTC")).toLocalDateTime()
        ).isEqualTo(LocalDateTime.of(2020, 7, 4, 23, 48, 0, 0));
    }
}

File: src/aws/lambdas/incremental_distribution/src/test/java/uk/nhs/nhsx/diagnosiskeydist/DistributionServiceWindowTest.kt
package uk.nhs.nhsx.diagnosiskeydist

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.CsvSource
import java.time.Instant


class DistributionServiceWindowTest {

    @ParameterizedTest
    @CsvSource(value = [
        "2020-07-04T11:45:59.999Z,early start date,false",
        "2020-07-04T11:48:00.000Z,late start date,false",
        "2020-07-04T12:46:00.000Z,even hour early start date,false",
        "2020-07-04T12:47:00.000Z,even hour late start date,false",
        "2020-07-04T11:46:00.000Z,early start date,true",
        "2020-07-04T11:47:59.999Z,late start date,true",
    ])
    fun `determines distribution start time window`(input: String, description: String, expected: Boolean) {
        assertThat(DistributionServiceWindow(input.asInstant()).isValidBatchStartDate)
            .describedAs("$description: $input should have invalid start time window")
            .isEqualTo(expected)
    }

    @ParameterizedTest
    @CsvSource(value = [
        "2020-07-04T00:00:00.000Z,2020-07-04T02:00:00.00Z",
        "2020-07-04T01:00:00.000Z,2020-07-04T02:00:00.00Z",
        "2020-07-04T01:59:59.999Z,2020-07-04T02:00:00.00Z",
        "2020-07-04T22:00:00.000Z,2020-07-05T00:00:00.00Z",
        "2020-07-04T23:00:00.000Z,2020-07-05T00:00:00.00Z",
        "2020-07-04T23:59:59.999Z,2020-07-05T00:00:00.00Z",
    ])
    fun `calculates next window`(input: String, expected: String) {
        val nextWindow = DistributionServiceWindow(input.asInstant())
            .nextWindow()

        assertThat(nextWindow)
            .isEqualTo(expected.asInstant())
    }

    @ParameterizedTest
    @CsvSource(value = [
        "2020-07-04T00:00:00.000Z,2020-07-04T04:00:00.00Z",
        "2020-07-04T01:00:00.000Z,2020-07-04T04:00:00.00Z",
        "2020-07-04T01:59:59.999Z,2020-07-04T04:00:00.00Z",
        "2020-07-04T22:00:00.000Z,2020-07-05T02:00:00.00Z",
        "2020-07-04T23:00:00.000Z,2020-07-05T02:00:00.00Z",
        "2020-07-04T23:59:59.999Z,2020-07-05T02:00:00.00Z",
    ])
    fun `calculates zip extraction exclusive time`(input: String, expected: String) {
        val zipExpirationExclusive = DistributionServiceWindow(input.asInstant())
            .zipExpirationExclusive()

        assertThat(zipExpirationExclusive)
            .isEqualTo(expected.asInstant())
    }

    @ParameterizedTest
    @CsvSource(value = [
        "2020-07-04T00:00:00.000Z,2020-07-04T01:46:00.00Z",
        "2020-07-04T01:00:00.000Z,2020-07-04T01:46:00.00Z",
        "2020-07-04T01:59:59.999Z,2020-07-04T01:46:00.00Z",
        "2020-07-04T22:00:00.000Z,2020-07-04T23:46:00.00Z",
        "2020-07-04T23:00:00.000Z,2020-07-04T23:46:00.00Z",
        "2020-07-04T23:59:59.999Z,2020-07-04T23:46:00.00Z",
    ])
    fun `calculates earliest batch start date within hour (inclusive)`(input: String, expected: String) {
        val withinHourInclusive = DistributionServiceWindow(input.asInstant())
            .earliestBatchStartDateWithinHourInclusive()

        assertThat(withinHourInclusive)
            .isEqualTo(expected.asInstant())
    }

    @ParameterizedTest
    @CsvSource(value = [
        "2020-07-04T00:00:00.000Z,2020-07-04T01:48:00.00Z",
        "2020-07-04T01:00:00.000Z,2020-07-04T01:48:00.00Z",
        "2020-07-04T01:59:59.999Z,2020-07-04T01:48:00.00Z",
        "2020-07-04T22:00:00.000Z,2020-07-04T23:48:00.00Z",
        "2020-07-04T23:00:00.000Z,2020-07-04T23:48:00.00Z",
        "2020-07-04T23:59:59.999Z,2020-07-04T23:48:00.00Z",
    ])
    fun `calculates latest batch start date within hour (exclusive)`(input: String, expected: String) {
        val withinHourExclusive = DistributionServiceWindow(input.asInstant())
            .latestBatchStartDateWithinHourExclusive()

        assertThat(withinHourExclusive)
            .isEqualTo(expected.asInstant())
    }

    private fun String.asInstant() = Instant.parse(this)
}
